README.md
# generator-semantic-release [![NPM version][npm-image]][npm-url] [![Build Status][travis-image]][travis-url] [![semantic-release](https://img.shields.io/badge/%20%20%F0%9F%93%A6%F0%9F%9A%80-semantic--release-e10079.svg)](https://github.com/semantic-release/semantic-release)

> A simple semantic release composable generator

## Installation

First, install [Yeoman](http://yeoman.io) and generator-semantic-release using
[npm](https://www.npmjs.com/) (we assume you have pre-installed
[node.js](https://nodejs.org/)).

```bash
npm install -g yo
npm install -g @goraxe/generator-semantic-release
```

Then generate your new project:

```bash
yo semantic-release
```

## Getting To Know Yeoman

- Yeoman has a heart of gold.
- Yeoman is a person with feelings and opinions, but is very easy to work with.
- Yeoman can be too opinionated at times but is easily convinced not to be.
- Feel free to [learn more about Yeoman](http://yeoman.io/).

## License

MIT © [goraxe]()

[npm-image]: https://badge.fury.io/js/@goraxe%2Fgenerator-semantic-release.svg
[npm-url]: https://npmjs.org/package/@goraxe/generator-semantic-release
[travis-image]: https://travis-ci.com/goraxe/generator-semantic-release.svg?branch=master
[travis-url]: https://travis-ci.com/goraxe/generator-semantic-release
[daviddm-image]: https://david-dm.org/goraxe/generator-semantic-release.svg?theme=shields.io
[daviddm-url]: https://david-dm.org/goraxe/generator-semantic-release


CHANGELOG.md
## [1.0.6](https://github.com/goraxe/generator-semantic-release/compare/v1.0.5...v1.0.6) (2023-12-18)


### Bug Fixes

* **deps:** update dependency chalk to v5 ([3811ed3](https://github.com/goraxe/generator-semantic-release/commit/3811ed35f983eaac317b1da2e0aa1aa5f244a34c))

## [1.0.5](https://github.com/goraxe/generator-semantic-release/compare/v1.0.4...v1.0.5) (2023-12-18)


### Bug Fixes

* **deps:** update dependency yosay to v3 ([78ff75f](https://github.com/goraxe/generator-semantic-release/commit/78ff75f420155f3770f9d058297d7222046fc795))

## [1.0.4](https://github.com/goraxe/generator-semantic-release/compare/v1.0.3...v1.0.4) (2023-12-17)


### Bug Fixes

* **renovate:** automerge github actions ([d2fa437](https://github.com/goraxe/generator-semantic-release/commit/d2fa4376430457743e238bc54db698dde6b24fa1))

## [1.0.3](https://github.com/goraxe/generator-semantic-release/compare/v1.0.2...v1.0.3) (2023-12-13)


### Bug Fixes

* **deps:** update dependency yeoman-generator to v4.13.0 ([8efd9c9](https://github.com/goraxe/generator-semantic-release/commit/8efd9c9767f5de4e943947e857b70798862af92d))

## [1.0.2](https://github.com/goraxe/generator-semantic-release/compare/v1.0.1...v1.0.2) (2020-03-14)


### Bug Fixes

* **security:** update dependencies ([91c0f42](https://github.com/goraxe/generator-semantic-release/commit/91c0f4296f45f33a31b71d6b64473e10ad129bb2))

## [1.0.1](https://github.com/goraxe/generator-semantic-release/compare/v1.0.0...v1.0.1) (2020-02-18)


### Bug Fixes

* public access ([167dd76](https://github.com/goraxe/generator-semantic-release/commit/167dd7668c7b1774fea4c79494262dc2d8c9360a))

# 1.0.0 (2020-02-18)


### Features

* add button to README.md ([87b466b](https://github.com/goraxe/generator-semantic-release/commit/87b466b98593714e961506e123e804969e475610))
* add sementic-release config file ([d0b57ff](https://github.com/goraxe/generator-semantic-release/commit/d0b57ffa53791562740a4f96d6755b893ff503e7))


__tests__/app.js
"use strict";
// Const path = require("path");
import path from "path";
import assert from "yeoman-assert";
import helpers from "yeoman-test";
import fs from "fs-extra";

// Const __dirname = url.fileURLToPath(new URL(".", import.meta.url));

import app from "../generators/app";

const badge =
  "[![semantic-release](https://img.shields.io/badge/%20%20%F0%9F%93%A6%F0%9F%9A%80-semantic--release-e10079.svg)](https://github.com/semantic-release/semantic-release)";

describe("generator-semantic-release:app", () => {
  /*  BeforeEach(async () => {
    runResult = await helpers.create("app", {},{})
  })
  */
  it("creates .releaserc", async () => {
    return helpers
      .create(app, {}, {})
      .run()
      .then(() => {
        assert.file([".releaserc"]);
        assert.jsonFileContent(".releaserc", {
          plugins: [
            "@semantic-release/commit-analyzer",
            "@semantic-release/release-notes-generator",
            "@semantic-release/changelog",
            "@semantic-release/git",
          ],
        });
      });
  });

  it("updates existing .releaserc", async () => {
    return helpers
      .create(app, {}, {})
      .withPrompts({})
      .inTmpDir((dir) => {
        fs.writeJson(path.join(dir, ".releaserc"), { branches: ["master"] });
      })
      .run()
      .then((runResult) => {
        runResult.assertFile(".releaserc");
        // Assert.file(path.join(dir, ".releaserc"));
        runResult.assertJsonFileContent(".releaserc", {
          branches: ["master"],
          plugins: [
            "@semantic-release/commit-analyzer",
            "@semantic-release/release-notes-generator",
            "@semantic-release/changelog",
            "@semantic-release/git",
          ],
        });
      });
  });

  describe("handles readme", () => {
    it("updates with a button when there is no button", async () => {
      return helpers
        .create(app, {}, {})
        .inTmpDir((dir) => {
          fs.writeFile(path.join(dir, "README.md"), "content");
        })
        .run()
        .then((runResult) => {
          /*
          RunResult.inTmpDir((dir) => {
            fs.writeFile(path.join(dir, "README.md"), "content");
          })
          */
          runResult.assertFile("README.md");
          runResult.assertFileContent("README.md", badge);
          runResult.assertFileContent("README.md", "content");
          //   });
        });
    });

    it("does not create a README.md", async () => {
      return helpers
        .create(app, {}, {})
        .run()
        .then((runResult) => {
          runResult.assertNoFile("README.md");
        });
    });

    it("leaves well alone if there is already a button", async () => {
      return helpers
        .create(app, {}, {})
        .inTmpDir((dir) => {
          fs.writeFile(path.join(dir, "README.md"), badge);
        })
        .run()
        .then((runResult) => {
          runResult.assertFile("README.md");
          runResult.assertEqualsFileContent("README.md", badge);
        });
    });
  });
});


generators/app/index.js
"use strict";
import Generator from "yeoman-generator";
import chalk from "chalk";
import yosay from "yosay";

const releasrc = {
  plugins: [
    "@semantic-release/commit-analyzer",
    "@semantic-release/release-notes-generator",
    "@semantic-release/changelog",
    "@semantic-release/git",
  ],
};

const badge =
  "[![semantic-release](https://img.shields.io/badge/%20%20%F0%9F%93%A6%F0%9F%9A%80-semantic--release-e10079.svg)](https://github.com/semantic-release/semantic-release)";

export default class extends Generator {
  prompting() {
    // Have Yeoman greet the user.
    this.log(
      yosay(
        `Welcome to the groovy ${chalk.red(
          "generator-semantic-release",
        )} generator!`,
      ),
    );

    const prompts = [];

    return this.prompt(prompts).then((props) => {
      // To access props later use this.props.someAnswer;
      this.props = props;
    });
  }

  writing() {
    if (this.fs.exists(this.destinationPath(".releaserc"))) {
      this.fs.extendJSON(this.destinationPath(".releaserc"), releasrc);
    } else {
      this.fs.writeJSON(this.destinationPath(".releaserc"), releasrc);
    }

    if (this.fs.exists(this.destinationPath("README.md"))) {
      let readme = this.fs.read(this.destinationPath("README.md"));
      if (!readme.includes(badge)) {
        readme = badge + "\n" + readme;
      }

      this.fs.write(this.destinationPath("README.md"), readme);
    }
  }

  end() {
    this.config.save();
  }
}
